### PR TITLE
Added support for dmaker.fan.p5 model

### PR DESCRIPTION
**Pull request by insajd**
p5 model does not report current direct_speed at which it's operating, so need to use Levels 1-4 which miio reports back to us.
Because p5 reporting only 4 levels, need to loop from 4th level back to 1st. That's the reason levels are different from za4 as well.
In order to support p5 and za4 simultaneously, added speed_list array.
According to my plan, it should work for za4 as before.

### CODE REVIEW

File: fan-xiaomi.js [hunk 1]
```diff
@@ -19,8 +19,29 @@ class FanXiaomi extends HTMLElement {
                 ui.querySelector('.var-title').textContent = this.config.name+' (Disconnected)';
                 return;
             }
+        }
+
+        const attrs = state.attributes;
+
+        let p5_speed_list = {"Level 1":1,"Level 2":35,"Level 3":70,"Level 4":100}
```
Comment (ikaruswill) [inline fan-xiaomi.js:26]:
```suggestion
        let p5_speed_list = [1, 35, 70, 100]
```
Perhaps using arrays, indexing them will be easier and more canonical later on in the speed selection, also mitigates typos

File: fan-xiaomi.js [hunk 2]
```diff
@@ -19,8 +19,29 @@ class FanXiaomi extends HTMLElement {
...
+        let za4_speed_list = {"Level 1":20,"Level 2":40,"Level 3":60,"Level 4":80, "Level 5":100}
```
Comment (ikaruswill) [inline fan-xiaomi.js:27]:
```suggestion
        let za4_speed_list = [20, 40, 60, 80, 100]
```

File: fan-xiaomi.js [hunk 3]
```diff
@@ -66,19 +86,24 @@ class FanXiaomi extends HTMLElement {
                     let icon = u.querySelector('.icon-waper > iron-icon')
                     let newSpeed
                     if (icon.getAttribute('icon') == "mdi:numeric-1-box-outline") {
-                        newSpeed = 40
+                        newSpeed = speed_list['Level 2']
```
Comment (ikaruswill) [inline fan-xiaomi.js:89]:
```suggestion
                        newSpeed = speed_list[1]
```
Arrays here in the speed selection

File: fan-xiaomi.js [hunk 4]
```diff
@@ -66,19 +86,24 @@ class FanXiaomi extends HTMLElement {
...
                     } else if (icon.getAttribute('icon') == "mdi:numeric-4-box-outline") {
-                        newSpeed = 100
-                        iconSpan.innerHTML = '<iron-icon icon="mdi:numeric-5-box-outline"></iron-icon>'
+                        if (model === 'dmaker.fan.p5') { // For p5 last speed is Level 4
+                            newSpeed = speed_list['Level 1']
+                            iconSpan.innerHTML = '<iron-icon icon="mdi:numeric-1-box-outline"></iron-icon>'
+                        } else {
+                            newSpeed = speed_list['Level 5']
+                            iconSpan.innerHTML = '<iron-icon icon="mdi:numeric-5-box-outline"></iron-icon>'
+                        }
```
Comment (ikaruswill) [inline fan-xiaomi.js:104]:
```suggestion
                        if speed_list[5] === undefined {
                            newSpeed = speed_list[0]
                            iconSpan.innerHTML = '<iron-icon icon="mdi:numeric-1-box-outline"></iron-icon>'
                        } else {
                            newSpeed = speed_list[4]
                            iconSpan.innerHTML = '<iron-icon icon="mdi:numeric-5-box-outline"></iron-icon>'
                        }
```
Using this should make the speed selection block model-agnostic

File: fan-xiaomi.js [hunk 5]
```diff
@@ -367,13 +435,23 @@ Natural
...
+        if (model === 'dmaker.fan.p5') { //p5 does not report direct_speed and natural_speed
+            direct_speed_int = speed_list[speed]
+            if (mode === 'nature') {
+                natural_speed = true
+            } else if (mode === 'normal') {
+                natural_speed = false
+            }
+        }
+        
+        if (direct_speed_int <= speed_list['Level 1']) {
```
Comment (ikaruswill) [inline fan-xiaomi.js:448]:
```suggestion
        if (direct_speed_int <= speed_list[0]) {
```
Using arrays for speed_list

**Comment by ikaruswill** [comment]
@insajd The p5 only has 4 levels of speed? That's interesting.. 

Are the speeds 1 35 70 100 arbitrary as well or are they given?

**Comment by insajd** [comment]
@ikaruswill I have merged your changes and tested on p5. Thanks!
Reagarding speed - yes, in Mi Home interface I can see only 4 levels.
I have taken speed list from https://github.com/syssi/xiaomi_fan/blob/93cd0ca0c91a955ae26c7af839536cac90b21619/custom_components/xiaomi_miio_fan/fan.py#L145
As I understand, their ranges are 
```
    FAN_SPEED_LEVEL1: range(1, 26),
    FAN_SPEED_LEVEL2: range(26, 51),
    FAN_SPEED_LEVEL3: range(51, 76),
    FAN_SPEED_LEVEL4: range(76, 101),
```
p5 does not report back direct_speed, just the level, so for speed 1 I can set anything from 1 till 25, and so on as long as they are in the speed level range - and all I'll know is that speed is Level 1.
Also, @syssi wrote here https://github.com/syssi/xiaomi_fan/issues/30#issuecomment-541293074 that he's planning on removing even those 4 speeds that we do have, and make it 3 in the future for Alexa/Google Home compatibility.
P.S. I  have messed up my PRs, as one PR is from branch, and another is from master. Hope you can manage this :)

**Comment by ikaruswill** [comment]
About the PR, no worries I got that, I've already merged the changes from that branch you based your changes on, to master, so now technically they're both `master` 😉

**Pull request merged**